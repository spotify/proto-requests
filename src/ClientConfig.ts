export class ClientConfig {
  keepCase: boolean;
  longs: string;
  enums: string;
  defaults: boolean;
  oneofs: boolean;
  rootLocation: string;
  async: boolean;

  constructor() {
    this.keepCase = false;
    this.longs = null;
    this.enums = null;
    this.defaults = false;
    this.oneofs = false;
    this.rootLocation = null;
    this.async = true;
  }
}
