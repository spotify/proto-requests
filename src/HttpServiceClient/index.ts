import { Method, Root, Service, Type} from 'protobufjs';
import { ClientConfig } from '../ClientConfig';
import { readFileSync } from 'fs';
import {
    getService,
    getTypes,
    getFullServiceName,
    getFullName,
} from '../utils/protobufUtils';
import {
    getTabsForDepth as gtd,
    getEntityNameFromNamespace,
    generateAnnotatedNamespaces,
    trimLeadingDot,
    generateNamespace,
} from "../utils/codeUtils";
import { fileHeader, polyfillFileHeader } from '../utils/fileHeader';
import { setupModule, declareObject, declareExport, exportDefault } from '../utils/moduleUtils';
import { startsWith } from "lodash";
import { WriteFunction, writeOutput, writeOutputSync } from '../utils/fileUtils';
import { Console } from '../types/console';

export interface CustomClass {
    filename: string;
    outputPath: string;
    fullName: string;
    contents?: string;
}

export type CustomClasses = Record<string, CustomClass>;

const CustomClasses: CustomClasses = {
    HttpServiceConfigOptions: {
        filename: require.resolve('./HttpServiceConfigOptions'),
        outputPath: 'protorequest/',
        fullName: 'protorequest.HttpServiceConfigOptions',
    },
};

const customResolvePath = (rootLocation: string, console: Console) => {
    return (origin: string, target: string): (string | null) => {
        console.log('customResolvePath rootLocation is: ', rootLocation);
        let result = "";
        if (target.startsWith(rootLocation)) {
            result = target;
        } else {
            result = `${rootLocation}${target}`;
        }
        console.log('customResolvePath returning: ', result);
        return result;
    };
};

export default class HttpServiceClient {
    root: Root;
    service: Service;
    methods: Method[];
    types: Type[];
    console: Console;

    constructor(
        file: string,
        options: ClientConfig,
        afterLoad: (error: Error, httpServiceClient: HttpServiceClient) => any,
        injectedConsole?: Console,
        ) {
        this.console = injectedConsole ? injectedConsole : console;
        const newRoot = new Root();
        newRoot.resolvePath = customResolvePath(options.rootLocation || '', this.console);
        newRoot.load(file, options, (err, root) => {
            if (err) {
                return afterLoad(err, null);
            }
            this.root = root;
            this.service = getService(this.root);
            this.types = getTypes(this.root);
            this.methods = this.service.methodsArray;
            return afterLoad(null, this);
        });
        return this;
    }

    getService = (): Service => {
        return getService(this.root);
    };

    generateCustomClasses = (): Record<string, CustomClass> => {
        const customClassesWithContents = {};
        Object.keys(CustomClasses).map((key) => {
            const filePath = CustomClasses[key].filename;
            const fileAsText = readFileSync(filePath).toString();
            const contents = `${polyfillFileHeader}
${fileAsText}`;
            customClassesWithContents[key] = { ...CustomClasses[key], contents };
        });
        return customClassesWithContents;
    };

    generateClient = (options): string => {
        const generatedClient = this.generateAnnotatedNamespacesForClient(options, this.root, '$root', 0);
        const serviceName = getService(this.root).name;
        return `
${fileHeader}

// Exported root namespace
${setupModule(options)}
${declareObject(options, "$root")}
${declareObject(options, serviceName)}
    
${generatedClient}

${exportDefault(options, '$root')}
${declareExport(options, "$root")}
${declareExport(options, serviceName)}
`;
    };

    generateProperties = (service: Service, methods: Method[], depth: number) => {
        return methods.map((method: Method) => {
            const trimmedMethodName = getEntityNameFromNamespace(method.name);
            return `${gtd(depth)}     * @property {function(${getFullName(this.root.lookupTypeOrEnum(method.requestType))}):Promise<${getFullName(this.root.lookupTypeOrEnum(method.responseType))}>} [${trimmedMethodName}] ${service.name} ${trimmedMethodName}`;
        }).join('\n');
    };

    generateService = (options, node, path, depth) => {
        const methodKeys = Object.keys(node.methods);
        const methodsArray = methodKeys.map(key => node.methods[key]);
        const generatedMethods = methodKeys.map(key => this.generateMethod(node.methods[key], node, path, options, depth)).join('');
        const properties = this.generateProperties(node, methodsArray, depth);
        const trimmedPath = trimLeadingDot(path);
        return `
${gtd(depth)}${getEntityNameFromNamespace(path)}.${node.name} = (function() {
${gtd(depth)}        
${gtd(depth)}    /**
${gtd(depth)}     * Properties of a ${node.name}.
${gtd(depth)}     * @memberof ${trimmedPath}
${gtd(depth)}     */
${gtd(depth)}     
${gtd(depth)}     /**
${gtd(depth)}      * Fetch function type, set when instantiating this service class.
${gtd(depth)}      * @callback FetchFunction
${gtd(depth)}      * @param {string} uri
${gtd(depth)}      * @param {Object} options
${gtd(depth)}      * @returns {Promise<any>}
${gtd(depth)}      */
${gtd(depth)}     
${gtd(depth)}    /**
${gtd(depth)}     * An HttpServiceClient class to send requests to the ${node.name}.
${gtd(depth)}     * @memberof ${trimmedPath}
${gtd(depth)}     * @classdesc ${node.name} to send requests to the ${node.name}.
${gtd(depth)}     * @param {FetchFunction} fetchFn
${gtd(depth)}     * @param {protorequest.HttpServiceConfigOptions} options
${gtd(depth)}     * @constructor
${gtd(depth)}     */
${gtd(depth)}    ${node.name} = function(fetchFn, options) {
${gtd(depth)}        this.fetchFn = fetchFn;
${gtd(depth)}        this.host = options.host;
${gtd(depth)}        this.baseUrl = options.baseUrl;
${gtd(depth)}        this.version = options.version || "";
${gtd(depth)}        this.headers = options.headers || new Headers();
${gtd(depth)}    };
${gtd(depth)}
${gtd(depth)}    /**
${gtd(depth)}     * Sends the fetch request decorated with the configured host.  Only for internal use.
${gtd(depth)}     */
${gtd(depth)}    ${node.name}.prototype.protoFetch = function(url, opts) {
${gtd(depth)}      return this.fetchFn(this.host + url, opts);
${gtd(depth)}    };
${gtd(depth)}
${gtd(depth)}    /**
${gtd(depth)}     * Sets the headers to be sent with all requests to this service.
${gtd(depth)}     * @function setHeaders
${gtd(depth)}     * @memberof ${trimmedPath}.${node.name}
${gtd(depth)}     * @param {Headers} headers
${gtd(depth)}     * @instance
${gtd(depth)}     * @returns {${trimmedPath}.${node.name}} ${node.name}
${gtd(depth)}     */
${gtd(depth)}    ${node.name}.prototype.setHeaders = function(headers) {
${gtd(depth)}      this.headers = headers;
${gtd(depth)}      return this;
${gtd(depth)}    };
${gtd(depth)}
${gtd(depth)}    /**
${gtd(depth)}     * Get the headers being sent with all requests to this service.
${gtd(depth)}     * @function getHeaders
${gtd(depth)}     * @memberof ${trimmedPath}.${node.name}
${gtd(depth)}     * @instance
${gtd(depth)}     * @returns {Headers}
${gtd(depth)}     */
${gtd(depth)}    ${node.name}.prototype.getHeaders = function() {
${gtd(depth)}      return this.headers;
${gtd(depth)}    };
${gtd(depth)}${generatedMethods}
${gtd(depth)}    return ${node.name};
${gtd(depth)}})();
`;
    };

    generateAnnotatedNamespacesForClient = (options, node, path, depth) => {
        const pathToLimitBy = `${getFullServiceName(this.root)}.`;
        const nodeHandlers = {
            service: this.generateService,
            enum: undefined,
            type:  undefined,
            field: undefined,
            method: undefined,
            root: undefined,
            namespace: (options, node, path, depth, nodeHandlers) => {
                // This code will keep us from publishing any empty namespaces that are referenced by the service's
                // proto file, but for whom we aren't generating any nodes.
                const fullPath = `${path}.${node.name}.`;
                const isARootWeShouldPublish = path === '$root' && startsWith(pathToLimitBy, `.${node.name}`);
                if (startsWith(pathToLimitBy, fullPath) || isARootWeShouldPublish) {
                    return generateNamespace(options, node, path, depth, nodeHandlers) || '';
                }
            },
        };
        return generateAnnotatedNamespaces(options, node, path, depth, nodeHandlers);
    };

    generateMethod = (method: Method, service: Service, namespace, options, depth) => {
        const trimmedNamespace = trimLeadingDot(namespace);
        return `
${gtd(depth)}    /**
${gtd(depth)}     * Calls ${service.name}.${method.name}
${gtd(depth)}     * @function ${method.name}
${gtd(depth)}     * @memberof ${trimmedNamespace}.${service.name}
${gtd(depth)}     * @param {${getFullName(this.root.lookupTypeOrEnum(method.requestType))}} request
${gtd(depth)}     * @instance
${gtd(depth)}     * @returns {Promise<${getFullName(this.root.lookupTypeOrEnum(method.responseType))}>} ${method.name}
${gtd(depth)}     */
${gtd(depth)}    ${service.name}.prototype.${method.name} = function(request) {
${gtd(depth)}        const url = '/' + this.baseUrl + '/${service.name}/' + this.version +  (this.version ? '/' : '') + '${method.name}';
${gtd(depth)}        const opts = {
${gtd(depth)}            method: 'POST',
${gtd(depth)}            body: JSON.stringify(request),
${gtd(depth)}            headers: this.headers
${gtd(depth)}        };
${gtd(depth)}        return this.protoFetch(url, opts)
${gtd(depth)}            .then(function(r) { return r.ok ? r.json() : Promise.reject(r); });
${gtd(depth)}    };
`;
    };
}
