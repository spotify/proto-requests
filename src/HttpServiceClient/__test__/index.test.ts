import { join } from 'path';
import { config as defaultConfig } from '../../../test/testconfig';
import { ClientConfig } from '../../ClientConfig';
import HttpServiceClient, { CustomClass, CustomClasses } from '../index';

let config;
try {
    config = require('../../../test/config').config;
} catch (ex) {
    console.log('no local config found, using testconfig.');
    config = defaultConfig;
}

const testFile = 'com/test/testservice.proto';

describe('HttpServiceClient', () => {
    let httpServiceClient: HttpServiceClient;
    let customClasses: CustomClasses = {};

    beforeAll(() => {
        const clientConfig = new ClientConfig();
        const mockConsole = {
            log: jest.fn()
        };
        clientConfig.keepCase = true;
        clientConfig.defaults = true;
        clientConfig.oneofs = true;
        clientConfig.rootLocation = join(__dirname, "/../../../proto/");
        clientConfig.async = true;

        return new Promise((resolve, reject) => {
            new HttpServiceClient(testFile, clientConfig, (err, client) => {
                if (err) {
                    const error = `Error loading proto files:\n${err}`;
                    console.error(error);
                    reject(error);
                    return;
                }

                customClasses = client.generateCustomClasses();
                httpServiceClient = client;
                resolve();
            }, mockConsole);
        });
    });

    it('testservice client snapshot', () => {
        const fileContents = httpServiceClient.generateClient(config);
        expect(fileContents).toMatchSnapshot();
    });

    it('testservice custom classes snapshot', () => {
        Object.keys(customClasses).forEach((key: string) => {
            const fileContents = customClasses[key];
            expect(fileContents.contents).toMatchSnapshot(key);
        });
    });
});