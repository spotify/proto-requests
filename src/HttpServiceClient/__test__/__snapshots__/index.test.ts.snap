// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpServiceClient testservice client snapshot 1`] = `
"
// THIS FILE IS AUTO-GENERATED BY PROTO-REQUEST
/* eslint-disable */
\\"use strict\\";


// Exported root namespace
Object.defineProperty(exports, '__esModule', { value: true });
var $root = {};
var TestService = {};
    

$root.com = (function() {
    /**
     * Namespace com.
     * @exports com
     * @namespace
     */
    var com = {};

    com.test = (function() {
        /**
         * Namespace test.
         * @memberof com
         * @namespace
         */
        var test = {};

        test.TestService = (function() {
                
            /**
             * Properties of a TestService.
             * @memberof com.test
             */
             
             /**
              * Fetch function type, set when instantiating this service class.
              * @callback FetchFunction
              * @param {string} uri
              * @param {Object} options
              * @returns {Promise<any>}
              */
             
            /**
             * An HttpServiceClient class to send requests to the TestService.
             * @memberof com.test
             * @classdesc TestService to send requests to the TestService.
             * @param {FetchFunction} fetchFn
             * @param {protorequest.HttpServiceConfigOptions} options
             * @constructor
             */
            TestService = function(fetchFn, options) {
                this.fetchFn = fetchFn;
                this.host = options.host;
                this.baseUrl = options.baseUrl;
                this.version = options.version || \\"\\";
                this.headers = options.headers || new Headers();
            };
        
            /**
             * Sends the fetch request decorated with the configured host.  Only for internal use.
             */
            TestService.prototype.protoFetch = function(url, opts) {
              return this.fetchFn(this.host + url, opts);
            };
        
            /**
             * Sets the headers to be sent with all requests to this service.
             * @function setHeaders
             * @memberof com.test.TestService
             * @param {Headers} headers
             * @instance
             * @returns {com.test.TestService} TestService
             */
            TestService.prototype.setHeaders = function(headers) {
              this.headers = headers;
              return this;
            };
        
            /**
             * Get the headers being sent with all requests to this service.
             * @function getHeaders
             * @memberof com.test.TestService
             * @instance
             * @returns {Headers}
             */
            TestService.prototype.getHeaders = function() {
              return this.headers;
            };
        
            /**
             * Calls TestService.doSomething
             * @function doSomething
             * @memberof com.test.TestService
             * @param {google.protobuf.Empty} request
             * @instance
             * @returns {Promise<google.protobuf.Empty>} doSomething
             */
            TestService.prototype.doSomething = function(request) {
                const url = '/' + this.baseUrl + '/TestService/' + this.version +  (this.version ? '/' : '') + 'doSomething';
                const opts = {
                    method: 'POST',
                    body: JSON.stringify(request),
                    headers: this.headers
                };
                return this.protoFetch(url, opts)
                    .then(function(r) { return r.ok ? r.json() : Promise.reject(r); });
            };

            /**
             * Calls TestService.doSomethingElse
             * @function doSomethingElse
             * @memberof com.test.TestService
             * @param {com.test.Model} request
             * @instance
             * @returns {Promise<com.test.Result>} doSomethingElse
             */
            TestService.prototype.doSomethingElse = function(request) {
                const url = '/' + this.baseUrl + '/TestService/' + this.version +  (this.version ? '/' : '') + 'doSomethingElse';
                const opts = {
                    method: 'POST',
                    body: JSON.stringify(request),
                    headers: this.headers
                };
                return this.protoFetch(url, opts)
                    .then(function(r) { return r.ok ? r.json() : Promise.reject(r); });
            };

            return TestService;
        })();

        return test;
    })();

    return com;
})();


exports[\\"default\\"] = $root;
exports[\\"$root\\"] = $root;
exports[\\"TestService\\"] = TestService;
"
`;

exports[`HttpServiceClient testservice custom classes snapshot: HttpServiceConfigOptions 1`] = `
"// THIS FILE IS AUTO-GENERATED BY PROTO-REQUEST
/* eslint-disable */
Object.defineProperty(exports, '__esModule', { value: true });
var $root = {};
var HttpServiceConfigOptions = {};

$root.protorequest = (function() {
  /**
   * Namespace protorequest.
   * @exports protorequest
   * @namespace
   */
  var protorequest = {};

  protorequest.HttpServiceConfigOptions = (function() {
    /**
     * Properties of a HttpServiceConfigOptions.
     * @memberof protorequest
     * @interface IHttpServiceConfigOptions
     * @property {string|null} [host] HttpServiceConfigOptions host
     * @property {string|null} [setHost] HttpServiceConfigOptions setHost
     * @property {string|null} [baseUrl] HttpServiceConfigOptions baseUrl
     * @property {string|null} [setBaseUrl] HttpServiceConfigOptions setBaseUrl
     * @property {string|null} [version] HttpServiceConfigOptions version
     * @property {string|null} [setVersion] HttpServiceConfigOptions setVersion
     * @property {Headers|null} [headers] HttpServiceConfigOptions headers
     * @property {Headers|null} [setHeaders] HttpServiceConfigOptions setHeaders
     */

    /**
     * Constructs a new HttpServiceConfigOptions.
     * @memberof protorequest
     * @classdesc Represents a HttpServiceConfigOptions.
     * @implements IHttpServiceConfigOptions
     * @constructor
     * @param {protorequest.IHttpServiceConfigOptions=} [properties] Properties to set
     */
    function HttpServiceConfigOptions(properties) {
      if (properties)
        for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
          if (properties[keys[i]] != null)
            this[keys[i]] = properties[keys[i]];
    }

    /**
     * HttpServiceConfigOptions host.
     * @member {string} host
     * @memberof protorequest.HttpServiceConfigOptions
     * @instance
     */
    HttpServiceConfigOptions.prototype.host = '';

    /**
     * Sets host
     * @function setHost
     * @memberof protorequest.HttpServiceConfigOptions
     * @param {string} value
     * @instance
     * @returns {protorequest.HttpServiceConfigOptions} HttpServiceConfigOptions
     */
    HttpServiceConfigOptions.prototype.setHost = function(value) {
      this['host'] = value;
      return this;
    };

    /**
     * HttpServiceConfigOptions baseUrl.
     * @member {string} baseUrl
     * @memberof protorequest.HttpServiceConfigOptions
     * @instance
     */
    HttpServiceConfigOptions.prototype.baseUrl = '';

    /**
     * Sets baseUrl
     * @function setBaseUrl
     * @memberof protorequest.HttpServiceConfigOptions
     * @param {string} value
     * @instance
     * @returns {protorequest.HttpServiceConfigOptions} HttpServiceConfigOptions
     */
    HttpServiceConfigOptions.prototype.setBaseUrl = function(value) {
      this['baseUrl'] = value;
      return this;
    };

    /**
     * HttpServiceConfigOptions version.
     * @member {string} version
     * @memberof protorequest.HttpServiceConfigOptions
     * @instance
     */
    HttpServiceConfigOptions.prototype.version = '';

    /**
     * Sets version
     * @function setVersion
     * @memberof protorequest.HttpServiceConfigOptions
     * @param {string} value
     * @instance
     * @returns {protorequest.HttpServiceConfigOptions} HttpServiceConfigOptions
     */
    HttpServiceConfigOptions.prototype.setVersion = function(value) {
      this['version'] = value;
      return this;
    };

    /**
     * HttpServiceConfigOptions headers.
     * @member {Headers} headers
     * @memberof protorequest.HttpServiceConfigOptions
     * @instance
     */
    HttpServiceConfigOptions.prototype.headers = '';

    /**
     * Sets headers
     * @function setHeaders
     * @memberof protorequest.HttpServiceConfigOptions
     * @param {Headers} value
     * @instance
     * @returns {protorequest.HttpServiceConfigOptions} HttpServiceConfigOptions
     */
    HttpServiceConfigOptions.prototype.setHeaders = function(value) {
      this['headers'] = value;
      return this;
    };

    /**
     * Converts the HttpServiceConfigOptions to an object
     * @function toObject
     * @memberof protorequest.HttpServiceConfigOptions
     * @instance
     * @returns {object} object
     */
    HttpServiceConfigOptions.prototype.toObject = function() {
      return JSON.parse(JSON.stringify(this));
    };

    return HttpServiceConfigOptions;
  })();

  HttpServiceConfigOptions = protorequest.HttpServiceConfigOptions;

  return protorequest;

})();

exports[\\"$root\\"] = $root;
exports[\\"HttpServiceConfigOptions\\"] = HttpServiceConfigOptions;
exports[\\"default\\"] = $root;
"
`;
